amqp/src/main/scala/akka/stream/alpakka/amqp/impl/AmqpPublishFlowStage.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.amqp.impl
+
+import akka.annotation.InternalApi
+import akka.stream.alpakka.amqp.{AmqpWriteSettings, WriteMessage}
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import akka.stream._
+
+import scala.util.{Failure, Success, Try}
+
+/**
+ * Connects to an AMQP server upon materialization and sends incoming messages to the server.
+ * Each materialized flow will create one connection to the broker.
+ */
+@InternalApi
+private[amqp] final class AmqpPublishFlowStage[O](settings: AmqpWriteSettings)
+    extends GraphStage[FlowShape[(WriteMessage, O), O]] { stage =>
+
+  val in = Inlet[(WriteMessage, O)]("AmqpPublishConfirmFlow.in")
+  val out = Outlet[O]("AmqpPublishConfirmFlow.out")
+
+  override def shape: FlowShape[(WriteMessage, O), O] = FlowShape.of(in, out)
+
+  override protected def initialAttributes: Attributes =
+    Attributes
+      .name("AmqpPublishFlowStage")
+      .and(ActorAttributes.IODispatcher)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with AmqpConnectorLogic {
```

Comment:
Hi @jcare44 
I indeed like your implementation better than mine (and you also have updated the documentation which I didn't in #1584 )
However, small suggestion - I really thing that `AmqpConnectorLogic` has to be abstract class with constructor, not trait - it removes some code duplication. What do you think?

Reply:
Hi @fancywriter, thanks for your comments.
Hum yes it could be an abstract class, but I don't see much value to it. What would you mutualise in it? In your PR, I only saw the `override settings` moved as argument, and I don't see what more we could put in it :/
In the other hand, I think that it being a `trait` allows for more flexibility in the future, since you can compose it with other traits and still be open for an abstract class to be created if need be.